pilot/pkg/networking/plugin/plugin.go

```diff
@@ -47,7 +50,29 @@ const (
 )
 
 // ModelProtocolToListenerProtocol converts from a config.Protocol to its corresponding plugin.ListenerProtocol
-func ModelProtocolToListenerProtocol(protocol config.Protocol) ListenerProtocol {
+func ModelProtocolToListenerProtocol(node *model.Proxy, protocol config.Protocol) ListenerProtocol {
+	switch protocol {
+	case config.ProtocolHTTP, config.ProtocolHTTP2, config.ProtocolGRPC, config.ProtocolGRPCWeb:
+		return ListenerProtocolHTTP
+	case config.ProtocolTCP, config.ProtocolHTTPS, config.ProtocolTLS,
+		config.ProtocolMongo, config.ProtocolRedis, config.ProtocolMySQL:
+		return ListenerProtocolTCP
+	case config.ProtocolUnsupported:
+		if util.IsProxyVersionGE13(node) {
```

Comment:
I think we really need e2e tests to cover this new behavior, especially testing it together with other plugins (e.g. authentication and authorization) to make sure they will work nicely with each other. Could be added in follow-up PRs as this code-path is new.

Reply:
+1 strong need for e2e/integration tests for this type of feature

Reply:
Added e2e test. I am open to add more e2e test cases. BTW, I also have some unit tests for protocol sniffing.